### PR TITLE
#857 - Make the Jenkins WAR path configurable by the `JENKINS_WAR` env property

### PR DESCRIPTION
**Pull request by MRamonLeon**
It fixes issue #857 by defining the **JENKINS_WAR** env property and using it in install-plugins.sh.

Also included a new test: install-plugins/custom-war 

During the development I used test-install-plugins to save some time, I've added to the Makefile. If it's problematic I'm fine removing it.

I also added some documentation as comments in the **install-plugins.sh** script to ease the use.

@oleg-nenashev @varyvol @rsandell @raul-arabaolaza

### CODE REVIEW

File: install-plugins.sh [hunk 1]
```diff
@@ -4,12 +4,27 @@
 #
 # FROM jenkins
 # RUN install-plugins.sh docker-slaves github-branch-source
+#
+# Environment variables:
```
Comment (oleg-nenashev) [inline install-plugins.sh]: It is great to have it here, but should it be also in Markdown docs?
https://github.com/jenkinsci/docker#preinstalling-plugins

File: install-plugins.sh [hunk 2]
```diff
@@ -4,12 +4,27 @@
...
+# REF: directory with preinstalled plugins. Default: /usr/share/jenkins/ref/plugins
+# JENKINS_WAR: full path to the jenkins.war. Default: /usr/share/jenkins/jenkins.war
+# JENKINS_UC: url of the Update Center. Default: ""
+# JENKINS_UC_EXPERIMENTAL: url of the Experimental Update Center for "experimental" versions of plugins. Default: ""
```
Comment (varyvol) [inline install-plugins.sh]:
```suggestion
# JENKINS_UC_EXPERIMENTAL: url of the Experimental Update Center for experimental versions of plugins. Default: ""
```

File: install-plugins.sh [hunk 3]
```diff
@@ -4,12 +4,27 @@
...
+# JENKINS_INCREMENTALS_REPO_MIRROR: url of the incrementals repo mirror. Default: ""
+# JENKINS_UC_DOWNLOAD: downloado url of the Update Center. Default: JENKINS_UC/download
```
Comment (varyvol) [inline install-plugins.sh]:
```suggestion
# JENKINS_UC_DOWNLOAD: download url of the Update Center. Default: JENKINS_UC/download
```

File: tests/install-plugins.bats [hunk 4]
```diff
@@ -201,3 +201,15 @@ SUT_IMAGE=$(sut_image)
...
+
+@test "Use a custom jenkins.war" {
+  # Build the image using the right Dockerfile setting a new war with JENKINS_WAR env and with a weird plugin inside
+  run docker_build_child $SUT_IMAGE-install-plugins-custom-war $BATS_TEST_DIRNAME/install-plugins/custom-war --no-cache
+  assert_success
+  # Assert the weird plugin is there
+  assert_output --partial 'my-happy-plugin:1.1'
+}
+
+@test "clean work directory" {
```
Comment (varyvol) [inline tests/install-plugins.bats]: Isn't there some kind of @after?

File: tests/install-plugins/custom-war/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+WORKDIR /test-custom-dockerfile
+# Add there a new weird plugin to assert
+COPY --chown=jenkins:jenkins WEB-INF/ WEB-INF/
+
+USER root
+RUN chown jenkins:jenkins /test-custom-dockerfile
+USER jenkins
+
+# Copy the original jenkins.war to the custom location 
+RUN cp /usr/share/jenkins/jenkins.war /test-custom-dockerfile/my-custom-jenkins.war && \
```
Comment (varyvol) [inline tests/install-plugins/custom-war/Dockerfile]: You could use here $JENKINS_WAR I think.

File: tests/install-plugins.bats (hunk 4 above)
Comment (MRamonLeon) [reply tests/install-plugins.bats]: There is a _teardown_ method like the JUnit _after_ method, but it's executed after every test and there are some tests which should be run after the previous one without the cleaning. I could add this instruction to the test itself but I did on the same way of the previous existing tests.

Comment (varyvol) [reply tests/install-plugins.bats]: If it's done like that in other places and there is certainty that the "clean up" test will be the last one executed, I'm fine.

**Comment by MRamonLeon** [comment]
@oleg-nenashev Merged the sections, take a look

**Comment by MRamonLeon** [comment]
The tests are failing with remoting exceptions: 
```
[2019-08-14T14:05:50.880Z] Cannot contact ubuntu-jenkinsinfrab4b856: hudson.remoting.ChannelClosedException: Channel "unknown": Remote call on ubuntu-jenkinsinfrab4b856 failed. The channel is closing down or has closed down

[2019-08-14T14:32:06.409Z] Could not connect to ubuntu-jenkinsinfrab4b856 to send interrupt signal to process
```

**Comment by oleg-nenashev** [comment]
Tested it locally, should be fine to merge

**Comment by slide** [comment]
I did a `git bisect` to find out why the automated build is failing and it pointed to this PR.

**Comment by slide** [comment]
Specifically this test: https://github.com/jenkinsci/docker/blob/master/tests/install-plugins.bats#L171

**Pull request merged**